utils/timings.ml

```diff
@@ -0,0 +1,139 @@
+(***********************************************************************)
+(*                                                                     *)
+(*                                OCaml                                *)
+(*                                                                     *)
+(*                     Pierre Chambart, OCamlPro                       *)
+(*                                                                     *)
+(*  Copyright 2015 Institut National de Recherche en Informatique et   *)
+(*  en Automatique.  All rights reserved.  This file is distributed    *)
+(*  under the terms of the Q Public License version 1.0.               *)
+(*                                                                     *)
+(***********************************************************************)
+
+type file = string
+
+type source_provenance =
+  | File of file
+  | Pack of string
+  | Startup
+
+type compiler_pass =
+  | All
+  | Parsing of file
+  | Preprocessing of file
+  | Typing of file
+  | Transl of file
+  | Generate of file
+  | Assemble of file
+  | Clambda of file
+  | Cmm of file
+  | Compile_phrases of file
+  | Selection of source_provenance
+  | Comballoc of source_provenance
+  | CSE of source_provenance
+  | Liveness of source_provenance
+  | Deadcode of source_provenance
+  | Spill of source_provenance
+  | Split of source_provenance
+  | Regalloc of source_provenance
+  | Linearize of source_provenance
+  | Scheduling of source_provenance
+  | Emit of source_provenance
+
+let timings : (compiler_pass, float * float option) Hashtbl.t = Hashtbl.create 20
+let reset () = Hashtbl.clear timings
+
+let start pass =
+  (* Cannot assert it is not here: a source file can be compiled
+     multiple times on the same command line *)
+  (* assert(not (Hashtbl.mem timings pass)); *)
+  let time = Sys.time () in
+  Hashtbl.add timings pass (time, None)
+
+let stop pass =
+  assert(Hashtbl.mem timings pass);
+  let time = Sys.time () in
+  let (start, stop) = Hashtbl.find timings pass in
+  assert(stop = None);
+  Hashtbl.replace timings pass (start, Some (time -. start))
+
+let time pass f x =
+  start pass;
+  let r = f x in
+  stop pass;
+  r
+
+let restart pass =
+  let previous_duration =
+    match Hashtbl.find timings pass with
+    | exception Not_found -> 0.
+    | (_, Some duration) -> duration
+    | _, None -> assert false
```

Comment:
A minor stylistic peculiarity that I keep stumbling upon in OCaml code is that sometimes people avoid parenthesis inside tuple syntax, sometimes they don't, and sometimes they even mix both styles together, right on the neighboring lines. I wish people were more consistent about such matters. It almost hurts :-(